core/dictgen/src/rootcling_impl.cxx

```diff
@@ -2239,6 +2239,43 @@ static bool GenerateAllDict(TModuleGenerator &modGen, clang::CompilerInstance *c
 }
 
 ////////////////////////////////////////////////////////////////////////////////
+/// Includes all headers in the given module from this interpreter.
+static void IncludeModuleHeaders(TModuleGenerator &modGen, clang::Module *module, cling::Interpreter &interpreter)
+{
+   // Make a list of modules and submodules that we can check for headers.
+   // We use a SetVector to prevent an infinite loop in unlikely case the
+   // modules somehow are messed up and don't form a tree...
+   llvm::SetVector<clang::Module *> modules;
+   modules.insert(module);
+   for (size_t i = 0; i < modules.size(); ++i) {
+      clang::Module *M = modules[i];
+      for (clang::Module *subModule : M->submodules())
+         modules.insert(subModule);
+   }
+   // Now we include all header files from the previously collected modules.
+   for (clang::Module *module : modules) {
```

Comment:
I would like to do this the way clang is doing it which is by marking each header as a submodule and then let the visibility settings inside the module enforce this (e.g., the order doesn't matter, but clang ensures that each header first needs to import the submodules of the used headers, and if it doesn't it can't see the specific decls from this header even if they were parsed before and errors out with a nice error message that shows what include is missing).

This will either happen when we get the GenerateModuleAction from clang (which would still be the best solution) or until I get around and check out how to do this from rootcling.
I'll add a `FIXME:` that this is documented.